src/main/kotlin/com/hedvig/accountservice/schedulers/domain/sagas/SubscriptionScheduledSaga.kt

```diff
@@ -27,125 +29,169 @@ import java.time.ZoneId
 
-    private fun checkSubscriptionsOfPeriod(
-        event: ScheduleSubscriptionEvent,
-        period: YearMonth,
-        memberId: String,
-        checkAllPeriods: Boolean,
-        now: Instant
-    ) {
-
-        currentlyCheckingPeriod = period
-
-        if (!subscriptionAlreadyScheduled(memberId, period)) {
-            val productSubscriptionsOfPeriod = getProductSubscriptions(memberId, period)
-
-            if (productSubscriptionsOfPeriod.isNotEmpty()) {
-                productSubscriptionsOfPeriod.forEach { productSubscription ->
-                    commandGateway.sendAndWait<AddEntryCommand>(AddEntryCommand(memberId, Entry.from(event, period, productSubscription, now)))
-                    logger.info {
-                        "SUBSCRIPTION entry added to memberId=${event.memberId} " +
-                                "of productType=${productSubscription.productType} in ${period.month} ${period.year} " +
-                                "with amount=${productSubscription.subscription}"
-                    }
-                }
-            }
-            addSubscriptionScheduled(event.memberId, period, event.scheduledBy, now)
-        }
-
-        val previousPeriod = period.minusMonths(1)
-
-        if (shouldCheckPeriod(previousPeriod, memberId, checkAllPeriods)) {
-            checkSubscriptionsOfPeriod(event, previousPeriod, memberId, checkAllPeriods, now)
-        }
+    try {
+      checkSubscriptionsOfPeriod(event, currentPeriod, event.memberId, checkAllPeriods, now)
+    } catch (exception: Exception) {
+      commandGateway.sendAndWait<ScheduleSubscriptionFailedCommand>(
+        ScheduleSubscriptionFailedCommand(
+          memberId = event.memberId,
+          year = Year.of(currentlyCheckingPeriod.year),
+          month = currentlyCheckingPeriod.month,
+          scheduledBy = event.scheduledBy
+        )
+      )
+      logger.error { "Schedule subscription of PERIOD (${currentlyCheckingPeriod.month} ${currentlyCheckingPeriod.year}) failed for memberId=${event.memberId} (Message: ${exception.message})" }
+      return
     }
 
-    private fun shouldCheckPeriod(period: YearMonth, memberId: String, checkAllPeriods: Boolean = false): Boolean {
-        if (period.isBefore(firstPeriod)) {
-            return false
-        }
-        if (!subscriptionAlreadyScheduled(memberId, period)) {
-            return true
+    commandGateway.sendAndWait<SubscriptionScheduledCommand>(
+      SubscriptionScheduledCommand(
+        event.memberId,
+        event.scheduledBy
+      )
+    )
+  }
+
+  private fun checkSubscriptionsOfPeriod(
+    event: ScheduleSubscriptionEvent,
+    period: YearMonth,
+    memberId: String,
+    checkAllPeriods: Boolean,
+    now: Instant
+  ) {
+    currentlyCheckingPeriod = period
+
+    if (!subscriptionAlreadyScheduled(memberId, period)) {
+      val productSubscriptionsOfPeriod = getProductSubscriptions(memberId, period)
+
+      val discountOfPeriod = getBillingDiscount(memberId, period)
```

Comment:
Change to a list instead

```suggestion
      val discountsOfPeriod = getBillingDiscount(memberId, period)
```